s3-store/src/main.rs

```diff
@@ -0,0 +1,349 @@
+use anyhow::Context;
+use bao_tree::blake3;
+use clap::{Parser, Subcommand};
+use futures::{future, FutureExt};
+use indicatif::{
+    HumanBytes, HumanDuration, MultiProgress, ProgressBar, ProgressDrawTarget, ProgressStyle,
+};
+use iroh_bytes::{
+    provider::{self, handle_connection, EventSender},
+    BlobFormat,
+};
+use iroh_io::{AsyncSliceReaderExt, HttpAdapter};
+use iroh_net::{key::SecretKey, ticket::BlobTicket, MagicEndpoint, NodeAddr};
+use iroh_s3_bao_store::S3Store;
+use serde::Deserialize;
+use std::{
+    fmt::{Display, Formatter},
+    str::FromStr,
+    sync::Arc,
+    time::Duration,
+};
+use tokio_util::task::LocalPoolHandle;
+use url::Url;
+
+/// Send a file or directory between two machines, using blake3 verified streaming.
+///
+/// For all subcommands, you can specify a secret key using the IROH_SECRET
+/// environment variable. If you don't, a random one will be generated.
+///
+/// You can also specify a port for the magicsocket. If you don't, a random one
+/// will be chosen.
+#[derive(Parser, Debug)]
+pub struct Args {
+    #[clap(subcommand)]
+    pub command: Commands,
+}
+
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub enum Format {
+    #[default]
+    Hex,
+    Cid,
+}
+
+impl FromStr for Format {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_ascii_lowercase().as_str() {
+            "hex" => Ok(Format::Hex),
+            "cid" => Ok(Format::Cid),
+            _ => Err(anyhow::anyhow!("invalid format")),
+        }
+    }
+}
+
+impl Display for Format {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Format::Hex => write!(f, "hex"),
+            Format::Cid => write!(f, "cid"),
+        }
+    }
+}
+
+fn print_hash(hash: &blake3::Hash, format: Format) -> String {
+    match format {
+        Format::Hex => hash.to_hex().to_string(),
+        Format::Cid => hash.to_string(),
+    }
+}
+
+#[derive(Subcommand, Debug)]
+pub enum Commands {
```

Comment:
The help comments are not super clear to me. If I understand correctly, the difference is that `ServerUrl` will mirror things to the node and then serve them, whereas `ServeS3` only downloads to calculate the outboard, but does not persist the data on the local node? If this is correct then this should be part of the help text IMO.

Reply:
No, none of them downloads content. ServeUrl serves a list of URLs, ServeS3 gets an s3 bucket index, then constructs the URLs from that.